### PR TITLE
Stop zeroing huge memAllocBuf() buffers

### PR DESCRIPTION
**Pull request by yadij**
memAllocBuf() buffers smaller than 64KB were not zeroed before this
change. Larger buffers (a.k.a. "huge buffers") were zeroed with
xcalloc(). We believe it is safe to stop zeroing those huge buffers
because all known code that allocates huge buffers also allocates
smaller ones for the same purpose. If some of that code relied on
zeroing for years, we would expect to see problems with smaller buffers.

Removing xcalloc() allows removal of mem*Rigid(), memBufStats() and all
string pools API complications.

The cache manager mgr:mem report section for string
statistics is also removed.

### CODE REVIEW

**Comment by yadij** [comment]
Changes built on PR 1590 and just waiting on that PR to merge first.

File: src/tests/stub_libmem.cc
Comment (rousskov) [inline src/tests/stub_libmem.cc:47]: Why is it safe to stop zeroing this memory? Please adjust PR description to answer that question.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/tests/stub_libmem.cc
Comment (yadij) [reply src/tests/stub_libmem.cc:47]: done

**Review by rousskov — CHANGES_REQUESTED** [review]
If you are OK with my PR title/description adjustments, then we are almost done here! Please ping me after merging master with prerequisite PR changes into this PR branch.

File: src/mem/old_api.cc
Comment (yadij) [reply src/mem/old_api.cc:353]: On inspection this should not be using `memReallocBuf()` at all. It is the first allocation of the buffer. Switching to that and clearing it with `memset()` in the upcoming rebase.

File: src/tests/stub_libmem.cc
Comment (rousskov) [inline src/tests/stub_libmem.cc:45]: If this PR effectively makes the undocumented memAllocBuf() gross_size address _optional_, it should make the corresponding change in memReallocBuf() as well, to keep the two sister functions in sync: All buffers allocated using memAllocBuf() ought to remain compatible with memReallocBuf().

Please do not forget to update _all_ memReallocBuf() implementations, including the one in src/mem/minimal.cc.

File: src/mem/old_api.cc
Comment (rousskov) [inline src/mem/old_api.cc:200]: This PR makes gross_size parameter optional in src/mem/minimal.cc and src/tests/stub_libmem.cc, but the parameter is still required in this (primary) memAllocBuf() implementation. This PR also adds a memAllocBuf() caller that supplies nil gross_size and can crash this code.

Please adjust to keep all memAllocBuf() implementations in sync. For example, we can make the parameter optional in all memAllocBuf() implementations, including this one:

```suggestion
        HugeBufVolumeMeter += net_size;
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/tests/stub_libmem.cc
Comment (yadij) [reply src/tests/stub_libmem.cc:45]: It is **effectively** optional now, the callers have not changed their expectations nor requirements. Whenever it can be provided it should be provided.

The `memReallocBuf` suffers from the design flaw that looking up which pool is responsible for deallocating `oldbuf` requires knowing `oldbuf` size. That is passed using `gross_size` - so this parameter is **not** optional for `memReallocBuf()`.

Dropping `memReallocBuf()` entirely is in the planning for a followup PR as soon as these larger removals get out of the way.

Comment (rousskov) [reply src/tests/stub_libmem.cc:45]: > `gross_size` is **not** optional for `memReallocBuf()`.

Agreed. Sorry I missed that fact when writing my change request.

> Dropping `memReallocBuf()` entirely is in the planning for a followup PR

I am resolving my change request under the assumption that you will promptly post a followup PR that removes memReallocBuf() and, hence, address the problem this change request was dedicated to -- a new dangerous inconsistency between memAllocBuf() and memReallocBuf().

**Review by rousskov — APPROVED** [review]

**Review by kinkie — APPROVED** [review]
ship it!

**Pull request closed without merging**
